Fix build with Cabal 3.4

georgefst opened the pull request:
# Current status

(see edit history for further historical context - I edited down since this comment was getting unwieldy)

We all seem to be agreed that the most important thing initially is making `wxc` a system package, independent of Cabal or even Haskell. Old-style Cabal, with its global installs, just about worked for this, but it's a very weird fit for how Cabal works these days. Fundamentally, `wxc` was never a Haskell library - it was just convenient to install it via Hackage.  We've taken some good strides in this direction, and are beginning to create packages for Nixpkgs and the AUR, at least, and perhaps later Homebrew, Msys, Chocolatey, a Debian PPA etc. For now a user may have to install manually, via something like:
```sh
git clone https://github.com/wxHaskell/wxHaskell
cd wxhaskell/wxc
./generate-version-header.sh
mkdir build
cd build
cmake -DCMAKE_INSTALL_PREFIX=/usr ..
make install

# uninstall with `rm -rf /usr/include/wxc-1.0 /usr/lib/pkgconfig/wxc.pc /usr/lib/libwxc.so*`
```

Once `wxc` is installed, one can just add `wx` or `wxcore` to their `build-depends` like any other Haskell library.

## Non-Linux

It's pretty important that everything should work across at least the three major OSs, since cross-platform support is kind of the whole point of wxWidgets! I've done some experimentation here, but we could really do with help from people with more experience of building and distributing C libraries on those platforms.

### MacOS

I have `wxc` building, and at least some examples working, after:
- Installing wxWidgets and pkg-config. I used Nix, but I'm sure there are other ways.
- Setting an env var: `export PKG_CONFIG_PATH=/usr/local/lib/pkgconfig`.
- Changing line 327 in `wxc/src/cpp/eljprinting.cpp` from `return (void*)new wxPostScriptPrintNativeData();` to `return NULL;`.
- Using `-DCMAKE_INSTALL_PREFIX=/usr/local` instead of `/usr` in the `cmake` step.

### Windows

After opening the "Mingw haskell shell", and running `pacman -S gcc make cmake mingw64/mingw-w64-x86_64-wxWidgets`, plus extending `generate-version-header.sh` to allow version `3.0`, the build then fails quickly in the cmake step with:

```
CMake Error at /usr/share/cmake-3.21.2/Modules/FindwxWidgets.cmake:992 (file):
  file failed to open for reading (No such file or directory):

    /c/Users/gthomas/Documents/wxHaskell/wxc/C:/ghcup/msys64/mingw64/include/wx-3.0/wx/version.h
Call Stack (most recent call first):
  CMakeLists.txt:161 (find_package)
```

Which looks maybe like some sort of weird path-concatenation error? Unfortunately I'm not really a Windows user, and have little idea of quite how the Haskell toolchain with MinGW/Msys is all supposed to fit together. Hopefully somebody with more experience in that area can help out.

## Further TODOs
- [ ] Make sure the new `wxc` make build system is as functional as the old one. This will involve poring through 700 lines of Haskell, much of it written around a decade ago. And without any help from the authors, unless @jodonoghue is feeling particularly kind. On the plus side it is reasonably well-commented.
EDIT: I've skimmed it and I suspect that the only significant problems we might see are on Windows. Perhaps we can get away with just `generate-version-header.hs` on Mac/Linux, and convert the important parts of the old `Setup.hs` in to a Windows batch script.
- [ ] Avoid hardcoding any paths. See https://github.com/wxHaskell/wxHaskell/pull/40#discussion_r663369687.
- [ ] Fix broken sample programs: https://github.com/wxHaskell/wxHaskell/pull/40#issuecomment-873445433
- [ ] Look in to all compiler warnings.
- [ ] Get all this merged and uploaded to Hackage. Hopefully whoever controls this repository will drop in at some point. Otherwise we'll have to think about forking and convincing the Hackage trustees to give us the keys, which is always an unpleasant last resort.

georgefst committed:
Fix build with Cabal 3.4

georgefst committed:
Remove executables with missing sources

georgefst commented:
> wxc and wxdirect are both unusual in that they aren't actually used as Haskell dependencies, but as system-level ones, yet they're still built via Cabal.

I guess for that reason it makes sense to actually `cabal install wxdirect` to get the binary in to `~/.cabal/bin`.

But `cabal install wxc` complains because "it does not contain any executables". Would older versions of Cabal really go and put `libwxc.so` somewhere sensible? Weird.

pranaysashank commented:
I am working on it on macOS. Initially I got wxc working with cabal but then later decided to move it out of Cabal's setup and to have a make based build. So it is the users job to install wxc. I have many of the sample programs working with almost no tweaks. Let me push my changes somewhere .

pranaysashank commented:
> But cabal install wxc complains because "it does not contain any executables". Would older versions of Cabal really go and put libwxc.so somewhere sensible? Weird.

You should just do a cabal build, if you're working on latest cabal, it'll place the library in the appropriate build folder in dist-newstyle .

pranaysashank commented:
@georgefst I pushed my changes here https://github.com/pranaysashank/wxHaskell/tree/ps/wxc-make, only tested on my computer. It's a very rough draft and I don't think is backwards compatible (which I understand is important for this project). 

I think cabal has so many idiosyncrasies that it's not worth it to build `wxc` with it. One another change we can have is to add wxdirect to `build-tool-depends` of wxcore.cabal.

If you want to test my branch, you would need to install wxc
```bash
cd wxc
mkdir build
cd build
cmake -DCMAKE_INSTALL_PREFIX="<appropriatelocation>" ..
make install
```

and once that's done, add the headers directory to cabal.project for wxdirect to generate new modules while building wxcore. 

I am not familiar with make / CMake , so it's very likely what I wrote is terrible. I also have a shell.nix in both wxc and WxHaskell, that might be useful while developing.

pranaysashank commented:
I tried to use `extra-bundled-libraries` field in wxc.cabal, but it didn't work the way I wanted it to (or probably need to fix some more things)

georgefst commented:
> > But cabal install wxc complains because "it does not contain any executables". Would older versions of Cabal really go and put libwxc.so somewhere sensible? Weird.
> 
> You should just do a cabal build, if you're working on latest cabal, it'll place the library in the appropriate build folder in dist-newstyle .

I'm not finding that to be the case. It ends up at `./dist-newstyle/build/x86_64-linux/ghc-8.10.2/wxc-0.93.0.0/build/libwxc.so.0.93.0.0`, and then if I try `cabal build wxcore` I get the classic `* Missing (or bad) C library: wxc`. I don't see how this would work anyway if we were getting `wxc` from Hackage.

> Initially I got wxc working with cabal but then later decided to move it out of Cabal's setup and to have a make based build. So it is the users job to install wxc.

That seems very sensible. Perhaps pre-v2 Cabal was easier to abuse for this sort of thing. I suspect that if wxHaskell has a future, it involves packaging up `wxc` as something that exists outside of Cabal (maybe I should try and put together an AUR package...).Then `wxcore` and `wx` can be packaged as normal Haskell libraries which require `wxc` as a system dependency. As for `wxdirect`, well, I don't actually know what it does exactly (is it just required for *building* `wxcore`?), so I'm unsure which side of the fence it should be on.

georgefst commented:
> I think cabal has so many idiosyncrasies that it's not worth it to build `wxc` with it. One another change we can have is to add wxdirect to `build-tool-depends` of wxcore.cabal.

Good idea. That kind of answers my last question.

georgefst commented:
> I pushed my changes here https://github.com/pranaysashank/wxHaskell/tree/ps/wxc-make

Awesome! I'm running out of time today, but I'll take a good look at this.

> I don't think is backwards compatible (which I understand is important for this project).

Backwards compatible with old versions of `wxWidgets` or with old versions of GHC/Cabal etc.? If the latter, I wonder if maybe it's time for a clean break. This project has a lot of cruft, particularly on the build system side. And I think the vast majority of the Haskell world is on Cabal 3 and GHC > 8.6.

pranaysashank commented:
> and then if I try cabal build wxcore I get the classic * Missing (or bad) C library: wxc. 
Ah! yes, I forgot about that. You need to modify wxcore's Setup.hs, you need to add the wxc's build directory path in extraLibDirs of wxcore's Setup.hs. You can get the build path of wxc, by querying the installedPackages

pranaysashank commented:
> Backwards compatible with old versions of wxWidgets or with old versions of GHC/Cabal etc.?

I imagine it's the latter (not really sure), perhaps if we know how many are using wxHaskell as is it'll help us answer other questions.

georgefst committed:
Fix internal dependencies

`wxc` and `wxdirect` are not Haskell libraries, so they are no longer registered as such.

The `--wxc` flag to `wxdirect` is now used to point to the location of the header files. Unfortunately this is hardcoded for now in `wxcore`'s `Setup.hs`.

The intention is that `wxc` is installed as a system-level dependency, but we haven't determined how best to do this yet. For now, it is required to manually shuffle around some files:
- `cabal build wxc`
- `find . -name libwxc.so.0.93.0.0 -exec sudo cp {} /usr/lib/libwxc.so \;`
- `sudo cp -r wxc/src/include /usr/include/wxc`

georgefst commented on wxcore/Setup.hs:
```diff
@@ -39,21 +39,9 @@ wxcoreDirectoryQuoted  = "\"" ++ wxcoreDirectory ++ "\""
-                _  -> error "wxcInstallDir: I'm confused. I see more than one wxc include directory from the same package"
-        Unambiguous [] -> error "wxcInstallDir: Cabal says wxc is installed but gives no package info for it"
-        _ -> error "wxcInstallDir: Couldn't find wxc package in installed packages"
+wxcInstallDir lbi = pure "/usr/include/wxc/"
```
This needs fixing.

georgefst replied:
Maybe something like https://github.com/pranaysashank/wxHaskell/commit/f78d79eb869019731f8ed839f603003d4050137e#diff-26b6d3b6d64b3d830c56815fb3f6d4d88043dbe0d3336eb3fe807ad91d3eda81R44.

But that presumably only works because the path is hardcoded in the cabal file instead.

pranaysashank replied:
One option we have is to set the build type to Configure and populate the fields in wxcore.buildinfo.in similar to how they do it for X11 Haskell bindings [X11.buildinfo.in](https://github.com/xmonad/X11/blob/master/X11.buildinfo.in)

georgefst replied:
That sounds promising! Not fully sure how it works though

fgaz replied:
Since it's a native library what about using pkg-config (through pkgconfig-depends) instead?

amigalemming replied:
> Since it's a native library what about using pkg-config (through pkgconfig-depends) instead?

This would be optimal, but e.g. on Debian the developer package of wx does not ship a .pc file:

~~~~
$ dpkg --listfiles libwxgtk3.0-gtk3-dev | fgrep pkg
~~~~

fgaz replied:
I mean for wxc (see the commented line)

georgefst replied:
FWIW, I think the question of how to distribute `wxc` is the only real blocker on this PR. I've been intending for a while to try packaging it for AUR and/or Nix. But really, I've just been hoping that someone who knows more about building C libraries comes along and offers to work on it.

fgaz replied:
I have a draft of a nix package, I can add it to nixpkgs.

If I try to build wxcore with it though I get this error:

```
src/haskell/Graphics/UI/WXCore/WxcTypes.hs:123:2: error:
     fatal error: wxc/wxc_def.h: No such file or directory
      123 | 
          |  ^              
    |
123 | 
    |  ^
```

fgaz replied:
Duh, there's wxc/generate-version-header.sh. Let's see if it works now...

fgaz replied:
Prerequisite for nix packaging: https://github.com/georgefst/wxHaskell/pull/1

georgefst replied:
> I have a draft of a nix package

Could you please share this?

fgaz replied:
[Here you go](https://github.com/fgaz/nixpkgs/blob/wxc/init/pkgs/development/libraries/wxc/default.nix). I'd like to implement [this](https://github.com/georgefst/wxHaskell/pull/2#discussion_r1204312467) before submitting it to nixpkgs though.

Maybe In the meantime I can add it to the wxHaskell repo

fgaz replied:
https://github.com/georgefst/wxHaskell/pull/5

georgefst committed:
Small fixes

Should maybe squash with previous commit.

georgefst commented:
> If you want to test my branch, you would need to install wxc
> 
> ```shell
> cd wxc
> mkdir build
> cd build
> cmake -DCMAKE_INSTALL_PREFIX="<appropriatelocation>" ..
> make install
> ```

This doesn't work for me, with `cmake version 3.20.3` on Manjaro Linux:
```
CMake Warning:
  No source or binary directory provided.  Both will be assumed to be the
  same as the current working directory, but note that this warning will
  become a fatal error in future CMake releases.


CMake Error: The source directory "/home/gthomas/Code/wxHaskell/wxc/build" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.
```

georgefst committed:
Move wxc to a make based build.

pranaysashank commented:
@georgefst Do you have the CMakeLists.txt file in wxc folder? Also it is `..` as in the parent directory at the end of the command `cmake -DCMAKE_INSTALL_PREFIX="<appropriatelocation>" ..` 

Other than any possible platform specific changes I think it should work ?

georgefst commented:
Argh, I was being an idiot. I was in `wxc/build`.

georgefst commented:
See my edits to the OP for a summary of where we're up to. Feels like good progress but that's as much time as I've got for this weekend at least.

pranaysashank commented:
I have these todos in my notes,
```
** TODO Fix TimeFlows & TimeFlowsEx sample program.
** TODO Fix TestTaskBarIcon sample program.
** TODO Fix PropertyGrid sample program.
** TODO Fix Print sample program
** TODO Fix text logger in ListCtrl sample
** TODO Fix CustomControl executable.
```

These are the programs that are either crashing or need some layout fixes on macOS. Once I am near a linux system, I may get time to update the status on there as well

georgefst committed:
Generate wxc_def.h

3f7650f fails on a clean build because this header is missing. So we extract the relevant part of the setup script we deleted in that commit, and turn it in to a Cabal script.

georgefst committed:
Simplify `wxc_def.h` generation

georgefst committed:
Simplify `wxc_def.h` generation further, probably breaking Windows compatibility

georgefst committed:
Give wxc build script better name

pranaysashank commented:
@georgefst I haven't added generating wxc_def.h logic to CMakeLists.txt because I am not sure what the exact requirement is. In CMakeLists.txt, the function find_package(wxWidgets) sets these variables,

```cmake
message(${wxWidgets_VERSION_STRING}) # 3.0.4
message(${wxWidgets_VERSION_MAJOR}) # 3
message(${wxWidgets_VERSION_MINOR}) # 0
message(${wxWidgets_VERSION_PATCH}) # 4
```

we can use the function `configure_file` to generate the header from a template (when you run `cmake ..`),
```cmake
configure_file(${CMAKE_SOURCE_DIR}/wxc_def.h.in ${CMAKE_SOURCE_DIR}/src/include/wxc_def.h @ONLY)
```

Even thought didn't test this on windows (yet!) one reason I used cmake is for its cross platform support, need to check if its working fine on windows.

pranaysashank commented:
I have added generation of header during cmake's configure step https://github.com/pranaysashank/wxHaskell/commit/659b5b03bd9796d03f127bda057319b93a732def , need to add compatibility check for wxWidgets version.

PS Right now it's generating the version number from the wxWidgets package found on the system, it's trivial to change it to the behaviour by the Setup file where we are generating header based on versions we mention as compatible.

georgefst committed:
Replace Haskell code for generating `wxc_def.h` with shell script

georgefst commented:
Cool. I think the shell script I just pushed mostly replicates the behaviour from the old `Setup.hs` (minus the caveat in the TODO comment). But it would be nicer to have it within cmake.

pranaysashank commented:
I added the check for compatibility into the cmake script https://github.com/pranaysashank/wxHaskell/commit/86521a801b8431184854b395eb66aa550d777715 . I think Setup.hs & your script works better when there are multiple versions of wxWidgets available. In the script I don't make an attempt to find a compatible version of package, but instead just check if the package we found is compatible.

See this note,

```cmake
# NOTE: find_package should try to find a wxWidgets version that is
# one of the ones in the list, starting from the highest first. That
# is not what happens currently, we check if the wxWidgets package we
# found is in one of the compatible versions.
set(compatibleVersions "3.0;2.9;")
```

georgefst commented:
Yeah, that's probably good enough. It seems unlikely many people would have multiple versions of wxWidgets installed. There are almost bound to be other regressions in this work anyway.

pranaysashank commented:
`cmake` is really something, spent many hours and couldn't get `wxc` to build (nixos). Thinking of some other approaches for `wxc` now

georgefst commented:
I'd like to have another crack at this over the holidays, but I can't even get `wxc` to build at all now:
```
CMake Error at /usr/share/cmake-3.22/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find wxWidgets (missing: wxWidgets_LIBRARIES)
Call Stack (most recent call first):
  /usr/share/cmake-3.22/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.22/Modules/FindwxWidgets.cmake:1025 (find_package_handle_standard_args)
  CMakeLists.txt:161 (find_package)
```

@pranaysashank Any ideas?

amigalemming commented:
> PS. are there particular reasons why:
> 
>     * the Github `Issues` pane is disabled for this repository?

I assume that the issues were still handled at SourceForge according to the package description:

http://sourceforge.net/p/wxhaskell/bugs/

amigalemming commented:
> I'd like to have another crack at this over the holidays, but I can't even get `wxc` to build at all now:

I got the same error.

I assume that it cannot be an error of uninstalled system libraries, because I could build wx with old Cabal-2.4 and GHC-8.0.2.

fgaz committed:
wxc: update to wx 3.2

georgefst committed:
Merge pull request #1 from fgaz/wx32

wxc: update to wx 3.2

fgaz committed:
wxc,wxcore: use pkgconfig

fgaz committed:
Fix some cabal warnings

georgefst committed:
Merge pull request #2 from fgaz/pkg-config

wxc,wxcore: use pkgconfig

georgefst committed:
Merge pull request #3 from fgaz/cabal-warnings

Fix some cabal warnings

fgaz committed:
Namespace headers under wxc/

fgaz committed:
Add nix shell

georgefst committed:
Merge pull request #5 from fgaz/nix-shell

Add nix shell

georgefst committed:
Merge pull request #4 from fgaz/wxc-subdirectory

Namespace headers under wxc/

fgaz committed:
Disable webview

We don't actually use it, and the default wxGTK nix package doesn't
provide it, so this commit fixes the nix shell too.
If we want to enable it again, we have to override wxGTK and set
withWebKit = true

fgaz commented:
Since the examples are building again and the owner of this repo seems to be unresponsive, I think it's time to find a new home for this project, so we can have an issue tracker, accept patches properly, and finally have a new release.

@georgefst @pranaysashank did you intend to maintain/host the project? If not I could do that. I don't promise any new features but at least I can keep it building and accept patches

georgefst committed:
Merge pull request #6 from fgaz/disable-webview

Disable webview

georgefst commented:
> Since the examples are building again and the owner of this repo seems to be unresponsive, I think it's time to find a new home for this project, so we can have an issue tracker, accept patches properly, and finally have a new release.

Fully agreed.

> @georgefst @pranaysashank did you intend to maintain/host the project? If not I could do that. I don't promise any new features but at least I can keep it building and accept patches

Given that I'm already a tad overwhelmed by my current OSS maintenance responsibilities, I'd be more than happy if you wanted to take that on! I'd continue to contribute where I can e.g. I'm currently in the process of creating an AUR package for wxc.

georgefst commented:
I managed to get in touch with a maintainer, who seems to be willing to give use the keys: https://sourceforge.net/p/wxhaskell/bugs/120/.

fgaz commented:
Thanks @georgefst! I replied to that thread.

By the way if we do end up merging this branch, I think we should remove the merge commits. We already have proper commit messages in the actual commits and the merge commits will be confusing if we end up hosting here or on another pr-based platform.

noughtmare commented:
I'm trying to install this on Windows, but I get:

```
Error: setup.exe: can't find source for Graphics\UI\WXCore\WxcClasses in
src/haskell
```

Ah, I guess I need to do something run `configure` configure somehow. I'll probably need to convert that to a batch script.

georgefst commented:
Yes, Windows has been a bit of an afterthought so far, because none of us are using it. Please post/contribute if you get it working!

noughtmare commented:
I have the wxcore Hello World example running on Windows 😄

noughtmare commented:
I believe the following reproduces what I did. Note that this requires GHC 9.4 or later (because that switches to the clang64 msys environment). And this assumes msys is installed by GHCup in the default location: C:\ghcup\msys64

```
git clone https://github.com/georgefst/wxHaskell.git
cd wxHaskell
$env:Path = 'C:\ghcup\msys64\bin;C:\ghcup\usr\bin;' + $env:Path
$env:MSYSTEM = 'CLANG64' # workaround
cd wxc
bash
pacman -S mingw-w64-clang-x86_64-{wxwidgets3.2-msw,cmake,ninja clang}
cmake -G Ninja . -DCMAKE_BUILD_TYPE=Release
cmake --build .
cmake --install . --prefix /clang64/
exit
cd ..
cabal build wxcore
```

pranaysashank commented:
Just wanted to give a small update. I re-visited this PR today and took a different path which I believe requires less effort and is not a drastic change from the current setup. In `wxc` package, currently all the `.cxx` files are declared in `extra-source-files` and built by running `gcc` in the custom Setup. We can achieve the same by simply declaring those files in `cxx-sources` instead. We'll still need either a custom setup or a configure setup to find the wxWidgets libraries from the output of `wx-config --libs all`.

I tested this approach with `cabal` 3.10.1.0 and `ghc` 9.4.3 and it works completely fine. With the new `system-cxx-std-lib`, I guess this should also be a reliable way to build on Windows. I'll open a new PR with my changes + changes from this PR (minux wxc as a separate library) tomorrow

amigalemming commented:
On Sat, 18 Nov 2023, Pranay Sashank wrote:

> Just wanted to give a small update. I re-visited this PR today and took 
> a different path which I believe requires less effort and is not a 
> drastic change from the current setup. In wxc package, currently all the 
> .cxx files are declared in extra-source-files and built by running gcc 
> in the custom Setup. We can achieve the same by simply declaring those 
> files in cxx-sources instead.

CXX-Sources is relatively new in Cabal. It was certainly not available at 
creation time of wxhaskell.

> We'll still need either a custom setup or a configure setup to find the 
> wxWidgets libraries from the output of wx-config --libs all.

Those custom setups make building generally more difficult and bound to 
specific Cabal versions. A different approach would be to let the user 
specify --extra-lib-dirs and --extra-include-dirs at the command-line, in 
a cabal.project.local or in .cabal/config, maybe in a Makefile that runs 
cabal with the result of wx-config.

pranaysashank commented:
> CXX-Sources is relatively new in Cabal. It was certainly not available at creation time of wxhaskell.

Of course, I am suggesting it as an alternative to what we decided to do earlier in this PR which is to create wxc as a separate C library and upload it to various distributions' package repositories. It didn't occur to me then to use cxx-sources

> Those custom setups make building generally more difficult and bound to specific Cabal versions. A different approach would be to let the user specify --extra-lib-dirs and --extra-include-dirs at the command-line, in a cabal.project.local or in .cabal/config, maybe in a Makefile that runs cabal with the result of wx-config.

I think it makes things as whole easier to use for the end user in this case. We can use a Configure based setup as well, if we want to avoid Custom setup.

georgefst commented:
This sounds like a great idea, as long as we can manage to use a configure-based setup rather than a full-blown custom one.

I still think distributing `wxc` far and wide as a system package is the right thing to do in the long term, but it's a lot of work, and this thread is already pretty old! I'm pleased to see anything that gets us closer to getting working versions of these packages back on Hackage.

pranaysashank commented:
I have pushed my changes to cxx-options branch, linking fails on windows so need to figure that one out. I checked that it builds on linux. I'll be looking into the Windows issues, if someone can check the following program on macOS, it'd be great.

```bash
cabal run samplesWxcore:BouncingBalls 
```
Make sure that `wx-config` program exists in PATH where you run cabal

pranaysashank commented:
Update: It works on Windows, I am able to run BouncingBalls executable and from cabal repl as well. Two things to remember:
1) Do not have two wxWidgets libraries, I had wxWidgets installed in mingw toolchain and clang64 tool chain. Setting `MSYSTEM=clang64` is not enough because ghc will still pass both library directories to the linker (in `-L` flags). If you do have them, you will see strange errors like symbol not found (even though you can clearly see them in the library) and you will waste a day on it
2) Use at least `cabal` version 3.8.1.0. Event.hs and Control.hs modules use `foreign import ccall "wrapper"` style code which creates `Event.o` & `Control.o` as archives with other object files in it and without this [fix](https://github.com/haskell/cabal/pull/8071) in 3.8.1.0 the `wxcore` library archive won't have symbols from those two modules. `ghcup` recommends cabal version 3.6, upgrade it to 3.8.1.0.

One downside is `cabal` invokes `ghc` to compile the `.cxx` files one at a time and that takes a really long time to build `wxc` library, I think it's an acceptable downside for now. I'll clean up the code and make a PR, testing on macos is still pending.

pranaysashank commented:
Update: I have a rudimentary CI setup going on `cxx-options`, I also added backwards compatibility till 3.0, since `ubuntu-latest` (22.04) still ships with wxWidgets 3.0.
There is an error on the macOS build:
```
     error: allocating an object of abstract class type 'wxPostScriptPrintNativeData'
```
 but from what I gather that error should happen on all other systems as well. However, I couldn't find any specific g++ / clang flag that would help me in either getting a warning or an error for that on linux.

I also added a flag `UseSystemWxc` for `wxcore` package, that when enabled, uses `wxc` from `pkg-config`.

There is a bug in programs compiled against `wxWidgets > 3.0`, I haven't yet got to its root cause but basically the following programs doesn't work as expected when compiled against later `wxWidgets` versions

```bash
cabal run samplesWx:BouncingBalls
```

This happens because, in `samples/wx/BouncingBalls.hs`:
```
       -- react on user input
       set p [on click         := dropBall vballs p             -- drop ball
             ,on clickRight    := (\_pt -> ballsFrame)          -- new window
             ,on (charKey 'p') := set t [enabled :~ not]        -- pause
             ,on (charKey '-') := set t [interval :~ \i -> i*2] -- increase interval
             ,on (charKey '+') := set t [interval :~ \i -> max 1 (i `div` 2)]  -- decrease interval
             ]
```
`on clickRight` overrides the previous `on click` event handlers, same with `charKey` event handler. Have to investigate what changed after 3.0 for this change in behaviour

fgaz commented:
I will be at Zurihac this year. Should we organize something? I created a [matrix room](https://matrix.to/#/#wxhaskell:matrix.org) if you'd like to discuss. I also migrated the repo to [codeberg](https://codeberg.org/wxHaskell/wxHaskell) so we can start merging stuff.

georgefst commented:
Yes, excellent idea! I booked my flights earlier this week, and this will be the first year that I won't be 24 hours late or stressing about giving a talk. I'll probably end up spending a lot of time hanging out in the HLS (/ general tooling) room if last year is anything to by, but I don't have any particular project ideas in mind. See you there!

It would be great to achieve something before this thread reaches its third birthday.

pranaysashank commented:
I spent a few hours today on this and pushed a commit to `cxx-options` branch that finally fixes event handling for wxWidgets > 3.0. It uses an undocumented function but I think we should be okay. 

We are now closer to making a release, here are the pending items:

1. `cabal run samplesWx:TimeFlows` throws a debug window, need to fix the invalid call
2. Most programs in `samplesWx` run almost fine, but some need slight tweaks
3. Check that the programs from [wiki](https://wiki.haskell.org/WxHaskell) can be built, patch them as needed (wxAsteroids, wxFruit, GeBoP)
4. Fix build on macOS (need someone with a mac)
5. Fix ci on windows, it used to build fine. Now we are getting the following [error](https://github.com/wxHaskell/wxHaskell/actions/runs/9118385974/job/25071425612)
```
Error: cabal-3.10.3.0.exe: The program 'ghc' version >=7.0.1 is required but
it could not be found.
```

If anyone wants to continue working on this, add `pkgs.callPackage ./wxc/package.nix {}` to shell.nix and 
```
package wxcore
  flags: +UseSystemWxc
```
to `cabal.project`. You can also let ghc build `wxc` instead of using from nix/system, but it's quite slow as ghc doesn't build c++ files in parallel

pranaysashank commented:
I have a strange error on my system, where clicking any of the print commands in
```bash
cabal run samplesWx:Print
```
crashes with `Segmentation fault (core dumped)` where as printing in
```bash
$(cabal list-bin samplesWx:Print)
```
runs perfectly fine

fgaz commented:
It works for me, I just get a bunch of warnings on the console:

```
(P:536631): Pango-CRITICAL **: 12:01:23.360: pango_font_description_set_size: assertion 'size >= 0' failed
```

pranaysashank commented:
strange, could be something about my environment. I am on nixos using a globally installed cabal & ghc.
